cuDSS/simple_matrix_market/matrix_market_reader.h

```diff
@@ -0,0 +1,197 @@
+        if (std::get<0>(a) == std::get<0>(b))
+            return std::get<1>(a) < std::get<1>(b);
+        return std::get<0>(a) < std::get<0>(b);
+    });
+
+
+    //Some checks
+    if ((found_lower) && (mview == CUDSS_MVIEW_UPPER)) {
+        if (verbose_err)
+            std::cerr << "\033[38;5;196m ERROR: mview is upper, but lower elements found \033[0m\n";
+        return MtxReaderErrorUpperViewButLowerFound;
+    }
+    if ((found_upper) && (mview == CUDSS_MVIEW_LOWER)) {
+        if (verbose_err)
+            std::cerr << "\033[38;5;196m ERROR: mview is lower, but upper elements found \033[0m\n";
+        return MtxReaderErrorLowerViewButUpperFound;
+    }
+    if (!(found_upper && found_lower) && mview == CUDSS_MVIEW_FULL) {
+        if (verbose)
+            std::cout << "\033[38;5;208m WARNING: mview is full, but only lower or upper elements found \033[0m\n";
+    }
+
+    //Initialize with 0's
+    std::fill(*csr_offsets_h, *csr_offsets_h + (n + 1), 0);
+
+    // First pass: count entries per row
+    for (auto& [i, j, val] : entries) {
+        if (i >= n || i < 0) {
+            if (verbose_err)
+                std::cerr << "\033[38;5;196m ERROR: Invalid row index " << i << " \033[0m\n";
+            return MtxReaderErrorOutOfBoundRowIndex;
+        }
+        if (j >= n || j < 0) {
+            if (verbose_err)
+                std::cerr << "\033[38;5;196m ERROR: Invalid col index " <<j << " \033[0m\n";
+            return MtxReaderErrorOfBoundColIndex;
+        }
+        (*csr_offsets_h)[i + 1]++;
+    }
+
+    // Prefix sum for empty rows
+    for (int i = 0; i < n; ++i) {
+        (*csr_offsets_h)[i + 1] += (*csr_offsets_h)[i];
+    }
+
+    // Second pass: fill in column and value arrays
+    std::vector<int> row_fill(n, 0);
+    for (auto& [i, j, val] : entries) {
+        int offset = (*csr_offsets_h)[i] + row_fill[i];
+        (*csr_columns_h)[offset] = j;
+        (*csr_values_h)[offset] = val;
+        row_fill[i]++;
+    }
+
+    // // Detect empty rows
+    for (int i = 0; i < n; ++i) {
+        if ((*csr_offsets_h)[i] == (*csr_offsets_h)[i + 1]) {
+            if (verbose)
+                std::cout << "\033[38;5;208m Warning: Row " << i << " is empty \033[0m\n";
+        }
+    }
+
+    if (verbose)
+        std::cout << "MATRIX READER: Completed with " << nnz << " nonzeros\n";
+    return 0;
+}
+
+template <typename _TYPE_>
+int rhs_reader(std::string filename, int& n, _TYPE_** b_values_h) {
+    std::ifstream file(filename);
+
+    if (!file.is_open()) {
+        std::cerr << "\033[38;5;196m Error: Could not open file \033[0m" << filename << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    std::string line;
+    bool foundSize = false;
+
+    int row = 0;
+    while (std::getline(file, line)) {
+        if (line.empty() || line[0] == '%') continue;
+
+        std::istringstream lineData(line);
+
+        if (!foundSize) {
+            int dummy0, dummy1;
+            lineData >> n >> dummy0 >> dummy1;
```

Comment:
The `array` Matrix Market format only lists `m n` for the size information in contrast to the `coordinate` format.

Note that you should verify that only a single right hand side is supplied, since you don't support more:
```suggestion
            int num_rhs;
            lineData >> n >> num_rhs;
            if (num_rhs != 1) {
                // fprintf or std::cerr the error that only a single RHS is supported
                return EXIT_FAILURE;
            }
```